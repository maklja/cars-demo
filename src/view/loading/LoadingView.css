.loading-spinner {
	font-size: 2em;
}

.loading-spinner.loading-spinner--hidden {
	visibility: hidden;
}

.loading-spinner .loading-spinner__text {
	text-align: center;
}
