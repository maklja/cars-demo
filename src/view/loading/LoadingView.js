import React, { Component } from 'react';
import PropTypes from 'prop-types';

import 'spinkit/css/spinners/8-circle.css'; // small library for spinner animations

import './LoadingView.css';

export default class LoadingSpinner extends Component {
	constructor(props) {
		super(props);

		this.state = {
			visible: props.visible
		};

		const spinnerCircleCount = 12; // 12 circles of spinner
		this._spinnerCircles = [];

		for (let i = 0; i < spinnerCircleCount; i++) {
			this._spinnerCircles.push(
				<div key={i} className={'sk-circle' + (i + 1) + ' sk-child'} />
			);
		}
	}

	componentDidMount() {
		const { delay } = this.props;
		// wait few sec before loading spinner is shown
		// some operations can last only few ms
		// in this case we don't what to "blink" spinner
		// on the screen and then hide it immediately
		// we use delay to controll this
		if (delay > 0) {
			this.visibleDelayTimeoutId = setTimeout(() => {
				this.setState({ visible: true });
			}, delay);
		}
	}

	componentWillUnmount() {
		// clear timeout on component unmount
		clearTimeout(this.visibleDelayTimeoutId);
	}

	render() {
		const { text } = this.props;
		const { visible } = this.state;
		return (
			<div
				className={`loading-spinner ${
					visible ? '' : 'loading-spinner--hidden'
				}`}
			>
				<div className="sk-circle">{this._spinnerCircles}</div>
				<div className="loading-spinner__text">
					<span>{text}</span>
				</div>
			</div>
		);
	}
}

LoadingSpinner.propTypes = {
	text: PropTypes.string,
	visible: PropTypes.bool,
	delay: PropTypes.number
};

LoadingSpinner.defaultProps = {
	text: '',
	visible: false,
	delay: 0
};

// create HOC for loading
export const withLoading = (
	Component, // React component
	{ delay = 2000, text = 'Loading data, please wait...', className = '' } = {} // options
) => {
	return props => {
		const { loading } = props;
		const view = loading ? (
			<div className={className}>
				<LoadingSpinner delay={delay} text={text} />
			</div>
		) : (
			<Component {...props} />
		);

		return view;
	};
};
