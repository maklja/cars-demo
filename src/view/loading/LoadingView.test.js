import React from 'react';
import renderer from 'react-test-renderer';

import LoadingView from './LoadingView';

it('LoadingView renders correctly with default props', () => {
	const tree = renderer.create(<LoadingView />).toJSON();
	expect(tree).toMatchSnapshot();
});

it('LoadingView renders correctly with custom text message', () => {
	const tree = renderer
		.create(<LoadingView visible={true} text="Loading message" />)
		.toJSON();
	expect(tree).toMatchSnapshot();
});

it('LoadingView renders correctly with delay and custom message', () => {
	jest.useFakeTimers(); // we need to controll setTimeout trigger inside LoadingView

	const loadingViewInstance = renderer.create(
		<LoadingView delay={10000} text="Loading message" />
	);

	// init state of instance will be { visible: false }
	const treeLoadingViewHidden = loadingViewInstance.toJSON();
	expect(treeLoadingViewHidden).toMatchSnapshot();

	jest.runOnlyPendingTimers(); // trigger timeout to make LoadingView component visible

	// check same instance with now new state { visible: true }
	const treeLoadingViewVisible = loadingViewInstance.toJSON();
	expect(treeLoadingViewVisible).toMatchSnapshot();
});
