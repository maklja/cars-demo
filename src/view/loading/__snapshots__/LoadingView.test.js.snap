// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoadingView renders correctly with custom text message 1`] = `
<div
  className="loading-spinner "
>
  <div
    className="sk-circle"
  >
    <div
      className="sk-circle1 sk-child"
    />
    <div
      className="sk-circle2 sk-child"
    />
    <div
      className="sk-circle3 sk-child"
    />
    <div
      className="sk-circle4 sk-child"
    />
    <div
      className="sk-circle5 sk-child"
    />
    <div
      className="sk-circle6 sk-child"
    />
    <div
      className="sk-circle7 sk-child"
    />
    <div
      className="sk-circle8 sk-child"
    />
    <div
      className="sk-circle9 sk-child"
    />
    <div
      className="sk-circle10 sk-child"
    />
    <div
      className="sk-circle11 sk-child"
    />
    <div
      className="sk-circle12 sk-child"
    />
  </div>
  <div
    className="loading-spinner__text"
  >
    <span>
      Loading message
    </span>
  </div>
</div>
`;

exports[`LoadingView renders correctly with default props 1`] = `
<div
  className="loading-spinner loading-spinner--hidden"
>
  <div
    className="sk-circle"
  >
    <div
      className="sk-circle1 sk-child"
    />
    <div
      className="sk-circle2 sk-child"
    />
    <div
      className="sk-circle3 sk-child"
    />
    <div
      className="sk-circle4 sk-child"
    />
    <div
      className="sk-circle5 sk-child"
    />
    <div
      className="sk-circle6 sk-child"
    />
    <div
      className="sk-circle7 sk-child"
    />
    <div
      className="sk-circle8 sk-child"
    />
    <div
      className="sk-circle9 sk-child"
    />
    <div
      className="sk-circle10 sk-child"
    />
    <div
      className="sk-circle11 sk-child"
    />
    <div
      className="sk-circle12 sk-child"
    />
  </div>
  <div
    className="loading-spinner__text"
  >
    <span>
      
    </span>
  </div>
</div>
`;

exports[`LoadingView renders correctly with delay and custom message 1`] = `
<div
  className="loading-spinner loading-spinner--hidden"
>
  <div
    className="sk-circle"
  >
    <div
      className="sk-circle1 sk-child"
    />
    <div
      className="sk-circle2 sk-child"
    />
    <div
      className="sk-circle3 sk-child"
    />
    <div
      className="sk-circle4 sk-child"
    />
    <div
      className="sk-circle5 sk-child"
    />
    <div
      className="sk-circle6 sk-child"
    />
    <div
      className="sk-circle7 sk-child"
    />
    <div
      className="sk-circle8 sk-child"
    />
    <div
      className="sk-circle9 sk-child"
    />
    <div
      className="sk-circle10 sk-child"
    />
    <div
      className="sk-circle11 sk-child"
    />
    <div
      className="sk-circle12 sk-child"
    />
  </div>
  <div
    className="loading-spinner__text"
  >
    <span>
      Loading message
    </span>
  </div>
</div>
`;

exports[`LoadingView renders correctly with delay and custom message 2`] = `
<div
  className="loading-spinner "
>
  <div
    className="sk-circle"
  >
    <div
      className="sk-circle1 sk-child"
    />
    <div
      className="sk-circle2 sk-child"
    />
    <div
      className="sk-circle3 sk-child"
    />
    <div
      className="sk-circle4 sk-child"
    />
    <div
      className="sk-circle5 sk-child"
    />
    <div
      className="sk-circle6 sk-child"
    />
    <div
      className="sk-circle7 sk-child"
    />
    <div
      className="sk-circle8 sk-child"
    />
    <div
      className="sk-circle9 sk-child"
    />
    <div
      className="sk-circle10 sk-child"
    />
    <div
      className="sk-circle11 sk-child"
    />
    <div
      className="sk-circle12 sk-child"
    />
  </div>
  <div
    className="loading-spinner__text"
  >
    <span>
      Loading message
    </span>
  </div>
</div>
`;
